Migrate RuntimeClass from a CRD to an internal API

tallclair opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

Migrate the RuntimeClass API to a built-in Kubernetes API (from a CRD). For background, see this SIG-Architecture thread: https://groups.google.com/d/topic/kubernetes-sig-architecture/RycGP4LvQZk/discussion

**Which issue(s) this PR fixes**:
For https://github.com/kubernetes/enhancements/issues/585

**Special notes for your reviewer**:

Only new non-boilerplate code is the validation (and associated test)

**Does this PR introduce a user-facing change?**:

```release-note
ACTION REQUIRED: The node.k8s.io API group and runtimeclasses.node.k8s.io resource have been migrated to a built-in API. If you were using RuntimeClasses (an default-disabled alpha feature, as of Kubernetes v1.12), then you must recreate all RuntimeClasses after upgrading, and the runtimeclasses.node.k8s.io CRD should be manually deleted. RuntimeClasses can no longer be created without a defined handler.

Introduce a RuntimeClass v1beta1 API. This new beta API renames `runtimeHandler` to `handler`, makes it a required field, and cuts out the spec (handler is a top-level field).
```

/milestone v1.14
/priority important-soon
/sig architecture node

/assign @thockin @vishh 
/cc @dchen1107 @derekwaynecarr

tallclair committed:
Migrate RuntimeClass to internal API

tallclair committed:
Migrate to internal node.k8s.io client

tallclair commented:
Fix registration.

tallclair commented:
I just realized this is missing all the REST handlers. Will update when this is ready for review.

roycaihw commented:
/cc @mbohlool

tallclair commented:
OK, I think this is ready for review now. Added in the rest handlers and some other boilerplate that was missed the first time around. Also verified that the CRD can safely be deleted after the internal version is installed.

tallclair commented:
Synced & regenerated code to fix verify errors.

tallclair committed:
Add node.k8s.io/v1beta1 API

tallclair commented:
Merged beta API into this PR after discussing w/ @thockin

vishh commented on pkg/apis/node/validation/validation.go:
```diff
@@ -0,0 +1,58 @@
+func ValidateRuntimeClassUpdate(new, old *node.RuntimeClass) field.ErrorList {
+	allErrs := genericvalidation.ValidateObjectMetaUpdate(&new.ObjectMeta, &old.ObjectMeta, field.NewPath("metadata"))
+
+	// All RuntimeClassSpec fields are immutable.
```
Assuming one of the future use cases is to explicitly specify resource overhead of containerizers, I can imagine such an overhead changing across different versions of containernerizers. Does this comment hold good or is the use case I'm referring to no longer in scope?

tallclair replied:
Yes, but all RuntimeClassSpec fields are **currently** immutable. I think we can just update this when we add a mutable field?

vishh replied:
Ok.

vishh commented on pkg/apis/node/types.go:
```diff
@@ -0,0 +1,71 @@
+	// If this is not specified (or empty), a default will be chosen by the
+	// implementation.
+	// +optional
+	RuntimeHandler *string
```
Validation seems to require this field to conform to DNS Subdomain (RFC 1123). Can you mention it here in comments?

tallclair replied:
Done.

vishh reviewed (commented):
@tallclair can you explain why the new `RuntimeClass` object cannot be an optional CRD? IIUC RuntimeClass was originally defined as a CRD and this PR now makes it a core object to circumvent CRD installation issues. My question is more around why does the installation issue matter if this object is `optional`?

tallclair commented on pkg/apis/node/types.go:
```diff
@@ -0,0 +1,71 @@
+
+// RuntimeClass defines a class of container runtime supported in the cluster.
+// The RuntimeClass is used to select or configure the container runtime that is
+// used to run all containers in a pod. RuntimeClasses are (currently) manually
```
It won't run until the matching class is declared. See https://github.com/kubernetes/kubernetes/blob/979bef156a8bfff112ab51292e38e036172d5f82/pkg/apis/core/types.go#L2636

tallclair commented on pkg/apis/node/types.go:
```diff
@@ -0,0 +1,71 @@
+	// values are specific to the node configuration & CRI implementation.  It is
+	// assumed that all handlers are available on every node, and handlers of the
+	// same name are equivalent on every node.
+	// For example, a handler called "runc" might specify that the runc OCI
```
>  ... and configuration that the CRI implementation **will use** ...

The point is that the CRI implementation is the component that actually acts on this value, not the kubelet.

Do you have any suggestions for how to describe this more clearly? Should I just drop the reference to the CRI implementation? I wanted to clarify that this isn't selecting between CRI-o and containerd though, but rather between runc and runsc, etc.

vishh replied:
What stops someone from using this field for a different purpose that what's intended coz it's upto the CRI runtime to interpret this field? Part of the confusion is due to the use of the word "runtime" we discussed in a previous comment.
Maybe you can drop "configuration" from `RuntimeHandler specifies the underlying runtime and configuration that the CRI implementation will use` to ensure this field is clearly expected to be used to specify the underlying `sandboxing` solution?
Re-reading the current version, I feel less confused. Maybe we can wait for @thockin to do a pass over this PR and see if he finds the comments unambiguous.

tallclair replied:
> What stops someone from using this field for a different purpose that what's intended coz it's upto the CRI runtime to interpret this field? 

Nothing, it could be completely abused. Maybe we'll need a concept of runtime conformance eventually?

> Maybe you can drop "configuration"

I don't want to exclude running the same "runtime binary" (e.g. runsc) with different configurations. For example, if we wanted to have a "gvisor" and "gvisor-lite" runtimeclass, where the "lite" version sacrificed some extra defenses in the name of performance (e.g. use the host TCP stack).

vishh replied:
> Nothing, it could be completely abused. Maybe we'll need a concept of runtime conformance eventually?

Hmm.. Probably not coz conformance would only make sense at the "Runtime Class" naming level right?

vishh replied:
Also, do you expect users to want to pass more arbitrary configuration information through Runtime Class object? The reason I ask is if the Runtime Handler fields should end up becoming an arbitrary list which is only interpretable by runtimes.

vishh replied:
nvm, I see below that you intend this not be used to pass arbitrary information by controlling the naming spec

tallclair commented on pkg/apis/node/types.go:
```diff
@@ -0,0 +1,71 @@
+	// If this is not specified (or empty), a default will be chosen by the
+	// implementation.
+	// +optional
+	RuntimeHandler *string
```
Yes. Currently that's equivalent to not specifying any runtime class, but in the future there may be more features in the spec that will make it more useful.

vishh replied:
Just for my edification, can you provide some examples? Underwhat scenarios would you not specify a runtime to be used, but specify other fields in this spec?

tallclair replied:
Once we have [pod overhead](https://docs.google.com/document/d/1EJKT4gyl58-kzt2bnwkv08MIUZ6lkDpXcxkHqCvvAp4/edit), you might want to define some overhead for the default runc runtime, but not bother updating all the nodes to define a "runc" handler (or to use pod overhead with a CRI implementation that doesn't support runtimeHandler, like dockershim).

vishh replied:
Got it, so essentially this field is optional for the sake of convenience.

tallclair commented on pkg/apis/node/types.go:
```diff
@@ -0,0 +1,71 @@
+// how the pod will be run. The RuntimeClassSpec should be treated as immutable.
+type RuntimeClassSpec struct {
+	// RuntimeHandler specifies the underlying runtime and configuration that the
+	// CRI implementation will use to handle pods of this class. The possible
```
All these terms are so overloaded. I like "containerizers", but my concerns are that no one will know what that means (not that anyone knows what runtime means either...). The field name is called "RuntimeHandler", so we're already locked into that to some degree (unless we want to rename it for beta? Although it's also called runtime_handler in the CRI).

Another possibility would be to say "OCI runtime" here, since that's what it is in practice. However, there's nothing to say I couldn't select between different non-OCI runtimes, or just different configurations of the CRI implementation.

Open to suggestions here...

vishh replied:
Does CRI "officially" support non-OCI runtimes? 
One alternative is to not use the word runtime and instead say this field is a config passed to CRI "runtimes" to choose appropriate "sandboxing" solutions. (Note "sandbox" is a new term in this case :( )

tallclair replied:
There are a couple of mentions of OCI in the [CRI spec](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto), but AFAIK it's not required. For instance, although gvisor & kata both support the OCI interface, they're working on integrating with containerd at the higher-level [contaired shim v2](https://github.com/containerd/containerd/issues/2426) API.

thockin replied:
This doesn't select between CRI implementations, right?  It passes info *through* CRI (ostensibly to "the runtime") to select a set of behaviors or sub-runtimes.  Am I getting that right?

I don't hate "runtime" but I do think it's worth contemplating

tallclair replied:
@thockin correct. This is essentially an opaque value from Kubernetes perspective. We pass it off to the CRI, and it is completely up to the CRI for what to do with it:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/cri/runtime/v1alpha2/api.proto#L365-L370

For both containerd & CRI-O, this is implemented as a key that selects from different OCI (or shim v2) level configurations.

An alternative approach to runtime_handler would be to just key off the RuntimeClass name directly, but I wanted to decouple the handler selection from the RuntimeClass resource. For instance, if an operator wanted to create a new RuntimeClass to handle rolling updates (as discussed [here](https://github.com/kubernetes/enhancements/pull/723)).

tallclair replied:
conclusion from offline discussion: if possible in CRD -> built in conversion, rename to `handler`

tallclair commented on test/e2e/node/runtimeclass.go:
```diff
@@ -92,54 +81,6 @@ var _ = SIGDescribe("RuntimeClass [Feature:RuntimeClass]", func() {
 		expectSandboxFailureEvent(f, pod, fmt.Sprintf("\"%s\" not found", rcName))
 	})
 
-	It("should recover when the RuntimeClass CRD is deleted [Slow]", func() {
```
It's no longer a CRD, so the test doesn't really make sense. I suppose we could keep it for the transition from CRD to internal, but maybe it would be better to just add an upgrade test in that case...

vishh replied:
The test title says `RuntimeClass CRD is deleted` and so I thought even the new core object could be deleted which would mean the test is still valid, Are you saying that the test servers a different purpose?

tallclair replied:
This was testing that the CRD itself (kind CustomResourceDefinition) was deleted, not that a RuntimeClass object was deleted. APIs can be disabled in the apiserver, but testing that would require some more specialized test infrastructure (since the apiserver would need to be restarted).

tallclair commented on pkg/apis/node/types.go:
```diff
@@ -0,0 +1,71 @@
+}
+
+// RuntimeClassSpec is a specification of a RuntimeClass. It contains parameters
+// that required to describe the RuntimeClass to the Container Runtime Interface
```
Done.

tallclair commented on pkg/apis/node/types.go:
```diff
@@ -0,0 +1,71 @@
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// RuntimeClass defines a class of container runtime supported in the cluster.
+// The RuntimeClass is used to select or configure the container runtime that is
```
I tried to clarify this, WDYT? The idea is that the RuntimeClass defines a "class of runtimes", but it's concrete usage in a cluster determines which runtime (that is part of that class) is run. The idea being that the runtime class could be "sandboxed", and the implementation could be gvisor in one cluster and kata containers in another. It remains to be seen how this will play out in practice...

vishh replied:
I like the current iteration better. It also looks like you expect some "common" Runtime Class names. May be you can mention that in the comments as well stating some common portable names are expected to be curated by the community over time?

tallclair replied:
That is mentioned in the [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/runtime-class.md#proposed-future-enhancements), but I'd rather not speculate about changes in the API comments.

tallclair commented:
Thanks for the review @vishh !  I assume it's fine to update the documentation on the alpha API as long as there aren't functional changes?

> @tallclair can you explain why the new `RuntimeClass` object cannot be an optional CRD? IIUC RuntimeClass was originally defined as a CRD and this PR now makes it a core object to circumvent CRD installation issues. My question is more around why does the installation issue matter if this object is `optional`?

I don't have a strong argument for why it shouldn't be a CRD, but this was the [decision that came out of SIG-Arch](https://groups.google.com/forum/#!topic/kubernetes-sig-architecture/RycGP4LvQZk/discussion). I think it boils down to 2 things: core non-alpha APIs should not depend on beta functionality, and if they ran into a missing feature or shortcoming it would cause problems later (or block the feature). Part of the idea is that it's OK to switch from CRD to internal going from alpha -> beta, but it would be problematic to go from CRD to internal in the beta -> GA transition. Feel free to way in on that [thread](https://groups.google.com/forum/#!topic/kubernetes-sig-architecture/RycGP4LvQZk/discussion) if you have more questions about it...

tallclair commented:
Question: should the old API be updated with the new comments? https://github.com/kubernetes/node-api/blob/master/pkg/apis/node/v1alpha1/types.go

vishh commented on pkg/apis/node/types.go:
```diff
@@ -0,0 +1,74 @@
+	// RuntimeHandler specifies the underlying runtime and configuration that the
+	// CRI implementation will use to handle pods of this class. The possible
+	// values are specific to the node configuration & CRI implementation.  It is
+	// assumed that all handlers are available on every node, and handlers of the
```
Are there no scenarios where runtime classes may be restricted to a subset of nodes, or is that left to the end user to manage via taints & tolerations?

tallclair replied:
At the moment it's left to the end user, but it's something I'm hoping to get inte v1.15. See https://docs.google.com/document/d/1W51yBNTvp0taeEss56GTk8jczqFJ2d6jBeN6sCSlYZU/edit for a few ideas around it.

vishh reviewed (commented):
> I assume it's fine to update the documentation on the alpha API as long as there aren't functional changes?

I'd think yes given there aren't much guarantees around alpha APIs.

> I don't have a strong argument for why it shouldn't be a CRD, but this was the [decision that came out of SIG-Arch](https://groups.google.com/forum/#!topic/kubernetes-sig-architecture/RycGP4LvQZk/discussion).

@thockin happened to share my concerns as well. I wonder if there was any misunderstanding around the impact of RuntimeClass on core k8s?

> core non-alpha APIs should not depend on beta functionality,

Are you referring to Pod Spec including a runtime class field in this case or something else?

> and if they ran into a missing feature or shortcoming it would cause problems later

Is this essentially a side-effect of no safe mechanism to extend core objects today?

> but it would be problematic to go from CRD to internal in the beta -> GA transition.

My question is why can't RuntimeClass object remain a CRD even post GA?

> should the old API be updated with the new comments?

I feel that will be generally useful for end users. Not sure if there is any official policy around it. @thockin ?

thockin commented on pkg/apis/node/types.go:
```diff
@@ -0,0 +1,74 @@
+	// containers in a pod.
+	// If this is not specified (or empty), a default will be chosen by the
+	// implementation.
+	// The RuntimeHandler must conform to the DNS Subdomain (RFC 1123)
```
Let's talk about contents.  What are the contents of this?  Where does a string that works here originate?  E.g. in storage class, the driver itself defines a string, which is expected to be a kubernetes qualified name (prefix.com/suffix) and is defined by the storage vendor.

If I were a cluster admin who wanted to use this, where would I get the value that I have to write here?  And is really anything allowed or should we require the same spec as storage(ish) ?

tallclair replied:
See my reply in the other thread, here: https://github.com/kubernetes/kubernetes/pull/74433#discussion_r262181978

The value is essentially opaque to kubernetes, and it's up to the CRI implementation for how to use it.

So why not just use a completely unconstrained string?

1. Much harder to tighten validation than relax it.
2. Bounding the values makes it easier on the CRI impls to not worry as much about formatting, and makes it more portable to different config setups (e.g. the key is both valid TOML and yaml, although the dots complicate it a bit.
3. Prevents this from becoming an arbitrary configuration mechanism (I don't want this to become a command-line string for launching the runtime).

We could use an even more strict requirement (e.g. `[a-zA-Z][-_a-zA-Z0-9]+[a-zA-Z0-9]`), but RFC 1123 is convenient and consistent with other Kubernetes names.

tallclair replied:
Conclusion from offline discussion: this field is opaque to kubernetes. It's up to the CRI implementation how to interpret it (e.g. if they want to namespace it). Action item is to clarify this in the doc comments.

thockin replied:
After IRL conversation: Document the hell out of this as "something the runtime understands, e.g. may be a key in a config file".   Does this need to be a subdomain, or is a DNS Label sufficient?  No dots...

tallclair replied:
Let's go with the stricter DNS label.

thockin commented on pkg/apis/node/types.go:
```diff
@@ -0,0 +1,74 @@
+	// For example, a handler called "runc" might specify that the runc OCI
+	// runtime (using native Linux containers) should be used to run the
+	// containers in a pod.
+	// If this is not specified (or empty), a default will be chosen by the
```
Let's talk about defaults.  This is different than storage and as such raises alarms.

The way storage works is that there's an admission controller that looks at all StorgaeClass resources for a label, and the one that is labelled is the default.  If a volume doesn't have a class defined, the admission will set the default.  This is nice because a) the admission could be replaced by something more policy-ish and b) it's changeable fairly trivially.

It has the side-effect that volumes which already exist do not get their class set, which is actually correct.

Now, pods that are already running *do* have a class, even if it is hidden today.  Until this struct carries parameters, that is, at which point you don't know if it matches the class in reality.

It would be more consistent to go the storage route.  Define admission and a label, decorate incoming pods.  Make this field required.

Why should I not require that?

tallclair replied:
> The way storage works is that there's an admission controller that looks at all StorgaeClass resources for a label, and the one that is labelled is the default.

This means that you can't have different defaults per namespace, and gets weird if you label multiple classes the default, or want to roll out a new default.

> If a volume doesn't have a class defined, the admission will set the default. This is nice because a) the admission could be replaced by something more policy-ish and b) it's changeable fairly trivially.

I am expecting that the default RuntimeClass will be a policy decision (e.g. via PodSecurityPolicy: https://github.com/kubernetes/kubernetes/pull/73795).

> It has the side-effect that volumes which already exist do not get their class set, which is actually correct.

Right, that's what this is addressing. If a pod does not have a runtimeclass (i.e. `pod.spec.runtimeClassName == nil`), then the CRI implementation has the final say. From the CRI perspective, that is equivalent to a pod using a RuntimeClass with a `nil` runtimeHandler. We could make runtimeHandler a required field in the RuntimeClass, but then there is no way for a RuntimeClass to target the backwards-compatible "legacy" handler. Being able to target that node-level default may be useful once we roll out other control-plane level features like PodOverhead & scheduling.

Does that make sense? If not, maybe we could grab a few minutes face-to-face.

thockin replied:
From IRL discussion, containerd (for example) has an unnamed default.  I'd like to request that containerd treat "" and "default" (or something) equivalently, and make this field required.

That has a side-effect that alpha has it as optional.  I think it should be OK to require this in beta as long as you satisfy the round-trip, which you'll need an annotation for.  E.g.

alpha { handler: "" } => beta { annotations: { "alpha-handler-was-empty": "true" }, handler: "default" }


@liggitt to see if he pukes on that idea.  I would never suggest it for anything but alpha compat.

tallclair replied:
Conclusion from offline discussion: make this field required, reserve a name for the behavior when no RuntimeClass is used.

Open questions:
- should this resolution of the reserved name be done by the kubelet (i.e. pass a nil value to the CRI), or should the resolution happen in the CRI itleself?
- what should the reserved value be? `default` or `legacy`?
- Is it ok to move this to a required field going from alpha to beta? What about going from CRD to bulit-in? @lavalamp @liggitt 

@kubernetes/sig-node-api-reviews

liggitt replied:
missing the offline context, but as far as mechanics are concerned, an optional field in alpha can easily be required in beta, as long as the alpha version has a default (defaults are per-version).

that looks like:
* alpha defaults
* beta doesn't default
* validation requires a value

tallclair replied:
Since the resources need to be recreated anyway for the CRD -> internal transition, is defaulting still needed? I was thinking of just making the field required in the built-in v1alpha1. Sure it's breaking, but so is moving to an internal API....

> reserve a name for the behavior when no RuntimeClass is used

Thinking about this a bit more, I think we should scrap this. Handler should be explicit.

thockin commented on pkg/apis/node/validation/validation.go:
```diff
@@ -0,0 +1,58 @@
+// ValidateRuntimeClassSpec validates the sink spec for audit
+func ValidateRuntimeClassSpec(spec *node.RuntimeClassSpec, fldPath *field.Path) field.ErrorList {
+	var allErrs field.ErrorList
+	if spec.RuntimeHandler != nil && *spec.RuntimeHandler != "" {
```
if it is truly optional, then "" should be invalid - it means someone went out of their way to specify the field with an invalid value.

OTOH, as per above, I don't think this should be optional, so "" is invalid by that metric, but would be field.Required() rather than field.Invalid()

tallclair replied:
I thought it was considered best-practice to treat empty & nil as equivalent, because json? I agree that if we decide runtimeHandler should be required, we should also forbid the empty value.

thockin replied:
nil and empty are distinct for strings, just not for lists (slices) and maps

thockin reviewed (commented):
I missed that sig-arch thread, so I apologize.  I do not intend to re-adjudicate it here, it's not worth the energy BUT...

If it was our intention to not rely on CRDs (beta) for other beta APIs, we did a poor job saying that.  I'm a little disappointed that we don't have the followthrough on this, but I do ACK that not all the pieces are in place for a smooth transfer.

tallclair commented on pkg/apis/node/types.go:
```diff
@@ -0,0 +1,74 @@
+
+	// Specification of the RuntimeClass
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#spec-and-status
+	Spec RuntimeClassSpec
```
Interesting, I didn't realize this was an option. I can't think of any use cases for a status right now, but I'm hesitant to completely rule it out. I could possible see a desire for some status around automated feature discovery of runtimes, or for like "availableVersions" (purely hypothetical). OTOH, dropping spec does make the API a bit neater, and is consistent with the other classes. WDYT?

thockin replied:
I like consistency

tallclair commented on pkg/apis/node/types.go:
```diff
@@ -0,0 +1,74 @@
+// RuntimeClassSpec is a specification of a RuntimeClass. It contains parameters
+// that are required to describe the RuntimeClass to the Container Runtime
+// Interface (CRI) implementation, as well as any other components that need to
+// understand how the pod will be run. The RuntimeClassSpec should be treated as
```
Done

tallclair commented:
Comments addressed & rebased. Changes include:

1. Remove spec from v1beta1
2. s/RuntimeHandler/Handler/
3. Handler is now an immutable, required DNS label formatted value (also applies to v1alpha1 RuntimeHandler)
4. Added details to Handler field comments
5. Added conversion.go and conversion_test.go for v1alpha1 API

PTAL

vishh commented on staging/src/k8s.io/api/node/v1alpha1/types.go:
```diff
@@ -0,0 +1,75 @@
+	// in a pod.
+	// The RuntimeHandler must conform to the DNS Label (RFC 1123) requirements
+	// and is immutable.
+	RuntimeHandler string `json:"runtimeHandler,omitempty" protobuf:"bytes,1,opt,name=runtimeHandler"`
```
Shouldn't the omitempty tag be dropped as well?

tallclair replied:
Good catch. Done.

tallclair commented on pkg/apis/node/v1alpha1/conversion_test.go:
```diff
@@ -0,0 +1,40 @@
+	node "k8s.io/kubernetes/pkg/apis/node"
+)
+
+func TestRuntimeClassConversion(t *testing.T) {
```
Oops! :flushed: Fixed.

vishh commented:
LGTM

liggitt commented on cmd/kube-apiserver/app/aggregator.go:
```diff
@@ -276,6 +276,8 @@ var apiVersionPriorities = map[schema.GroupVersion]priority{
 	{Group: "coordination.k8s.io", Version: "v1"}:               {group: 16500, version: 15},
 	{Group: "coordination.k8s.io", Version: "v1beta1"}:          {group: 16500, version: 9},
 	{Group: "auditregistration.k8s.io", Version: "v1alpha1"}:    {group: 16400, version: 1},
+	{Group: "node.k8s.io", Version: "v1alpha1"}:                 {group: 16300, version: 9},
```
make alpha=1, beta=9 to match others

tallclair replied:
Done (although 9 is actually more common for v1alpha1, 3 vs 2).

thockin reviewed (commented):
API LGTM
/approve

thockin commented:
Maybe squash commits? :)

tallclair commented:
Thanks! I'll clean up the commits & rebase.

tallclair committed:
generated files

tallclair commented:
Squashed, rebased, and ready to go!

vishh reviewed (approved):
/lgtm
/approve

tallclair commented:
/test pull-kubernetes-integration

tallclair commented:
/test pull-kubernetes-integration

tallclair commented:
rebased

vishh commented:
/lgtm (again)

spiffxp commented:
v1.14 release lead here, this is a size/XXL PR that is attempting to land right before Code Freeze... are you _really sure_ this is a good idea?

(I'm crossing my fingers and hoping the answer is yes, given the recent sustained effort and reviews, and the fact that this ties back to an enhancement slated for this release)

tallclair commented:
Functionally, this PR is relatively small. The size comes from all the generated files & boilerplate. I'd really like this to land for v1.14, to unblock RuntimeClass usage.

tallclair commented:
/retest

dchen1107 commented:
/lgtm
/approve
